Prepare the app for deployment on Heroku

gclssvglx opened the pull request:
Following the advice to [Set up Heroku review apps for pull requests](https://docs.publishing.service.gov.uk/manual/review-apps.html) and [Test & build a project with GitHub Actions](https://docs.publishing.service.gov.uk/manual/test-and-build-a-project-with-github-actions.html) - this change:

* Updates rubocop to use the `rubocop-govuk` gem - and corrects issues found 
* Adds a `Procfile`
* Updates the `Gemfile` with advice on keeping the Ruby version consistent 
* Adds an `app.json` file - this defines the project structure and any required Heroku add-ons
* Adds GitHub Actions to act as a CI layer with builds and tests the application
* Updates the production URL to `DATABASE_URL` for Heroku

[Trello](https://trello.com/c/gTnbKiF1/673-move-the-foreign-travel-deep-link-checker-application-to-alphagov-govuks-heroku)

gclssvglx committed:
Use rubocop-govuk and correct issues

gclssvglx committed:
Add Procfile

gclssvglx committed:
Update Gemfile to ensure Ruby version is correct

gclssvglx committed:
Add app.json

kevindew commented on Gemfile:
```diff
@@ -1,82 +1,40 @@
-
-gem 'httparty'
-gem 'nokogiri'
+ruby "2.7.2"
```
We tend to use: https://github.com/alphagov/smart-answers/blob/b2c4c330dbe128ade02600ebec5b0c3239ae9bca/Gemfile#L3 to save duplicating the .ruby-version file.

gclssvglx replied:
Thanks - have added this in this commit https://github.com/alphagov/foreign-travel-content-checker/pull/1/commits/6261f07bf6fd4a5f9617a4f56e31311f656638b3

kevindew commented on Gemfile:
```diff
@@ -1,82 +1,40 @@
+gem "pg", "~> 1.1"
+gem "puma", "~> 5.0"
+gem "rails", "~> 7.0.2", ">= 7.0.2.2"
+gem "redis", "~> 4.0"
```
Minor given the status of this repo, but these lines don't match GOV.UK conventions: https://docs.publishing.service.gov.uk/manual/conventions-for-rails-applications.html#gemfile-organisation

gclssvglx replied:
Thanks - fixed

gclssvglx committed:
Add production and test database URLs

kevindew commented on Procfile:
```diff
@@ -0,0 +1 @@
+web: bin/rails server -p $PORT -e $RAILS_ENV
```
Minor, the commit message for this is spelt wrong s/Profile/Procfile/

gclssvglx replied:
Thanks - fixed

kevindew commented on Gemfile:
```diff
@@ -3,7 +3,7 @@
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
-ruby "2.7.2"
+ruby File.read(".ruby-version").chomp
```
Nice 👍

kevindew commented on app.json:
```diff
@@ -0,0 +1,44 @@
+  "name": "foreign-travel-content-checker",
+  "repository": "https://github.com/alphagov/foreign-travel-content-checker",
+  "env": {
+    "GOVUK_APP_DOMAIN": {
+      "value": "www.gov.uk"
+    },
+    "GOVUK_WEBSITE_ROOT": {
+      "value": "https://www.gov.uk"
+    },
+    "PLEK_SERVICE_CONTENT_STORE_URI": {
+      "value": "https://www.gov.uk/api"
+    },
+    "PLEK_SERVICE_SEARCH_URI": {
+      "value": "https://www.gov.uk/api"
+    },
+    "PLEK_SERVICE_STATIC_URI": {
+      "value": "assets.digital.cabinet-office.gov.uk"
+    },
```
Do you know if you need all these env vars? they seem the type that are needed for GOV.UK dependencies that I doubt this app has.

gclssvglx replied:
Nope, the app doesn't need any of these. Just a copy-and-paste issue as I wasn't sure. Fixed.

kevindew commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,40 @@
+      - name: Setup the database
+        env:
+          RAILS_ENV: test
+          POSTGRES_USER: postgres
+          POSTGRES_PASSWORD: postgres
+          POSTGRES_DB: foreign_travel_content_checker_test
```
It'd be more common to set this up through a DATABASE_URL environment variable. 

Typically for testing apps we've used TEST_DATABASE_URL e.g: https://github.com/alphagov/govuk-docker/blob/main/projects/content-publisher/docker-compose.yml#L31-L33

gclssvglx replied:
Thanks - removed

gclssvglx replied:
When I tried to remove these, I got a `postgres service is unhealthy` error. Found this advice: https://github.com/jveillet/tentacles-dashboard/issues/205, then this https://github.community/t/mysql-service-never-comes-up-healthy-in-action/16890 which suggests that something has changed with GitHub actions and DRYing it no longer works 🤷

kevindew commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,40 @@
+      postgres:
+        image: postgres:latest
+        env:
+          POSTGRES_USER: postgres
+          POSTGRES_PASSWORD: postgres
+          POSTGRES_DB: foreign_travel_content_checker_test
```
I'm not sure you actually need much of this as per: https://docs.publishing.service.gov.uk/manual/test-and-build-a-project-with-github-actions.html#gov-uk-rails-application-with-postgres-redis-yarn-and-gov-uk-content-schemas-dependencies

gclssvglx replied:
Thanks - removed

gclssvglx replied:
See https://github.com/alphagov/foreign-travel-content-checker/pull/1#discussion_r826119433

kevindew commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,40 @@
+          POSTGRES_PASSWORD: postgres
+          POSTGRES_DB: foreign_travel_content_checker_test
+        run: |
+          bundle exec rails db:migrate
+          bundle exec rails db:seed
```
You'll probably want this to create the database too. You can do `db:reset` which will drop, create and seed.

It's not recommended to using db:migrate as a way to set-up a database. I've not found the authoritative documentation of this but this one at least sets the tone: https://edgeguides.rubyonrails.org/active_record_migrations.html#old-migrations

gclssvglx replied:
Thanks - fixed

kevindew commented on config/database.yml:
```diff
@@ -58,6 +58,9 @@ development:
   database: foreign_travel_content_checker_test
+  host: localhost
+  username: postgres
+  password: postgres
```
You should be able to remove this

gclssvglx replied:
Thanks - removed

gclssvglx replied:
See https://github.com/alphagov/foreign-travel-content-checker/pull/1#discussion_r826119433

kevindew commented on lib/tasks/security.rake:
```diff
@@ -2,5 +2,5 @@
 desc "Run Brakeman"
 task security: [:environment] do
-  sh "bundle exec brakeman"
+  sh "bundle exec brakeman . --except CheckRenderInline --no-summary --quiet"
```
Might want to put this in a separate commit and explain why these options exist.

kevindew reviewed (commented):
Lots of great additions here 👍  I've added a few comments to resolve.

gclssvglx committed:
Add GitHub Actions to act as a CI layer

gclssvglx committed:
Add security task to run brakeman

Runs brakeman without the normal verbose output (--quiet) and without showing
the overall findings summary (--no-summary)
